common/data_refinery_common/models/attributes.py

```diff
@@ -0,0 +1,90 @@
+from django.db import models
+
+from data_refinery_common.models.ontology_term import OntologyTerm
+
+BOOL = "bool"
+INT = "int"
+FLOAT = "float"
+ONTOLOGY_TERM = "ont"
+
+
+class AbstractAttribute(models.Model):
+    """This is an abstract class that defines all of the properties of a single
+    attribute on either a sample or an experiment. We then subclass this to
+    associate attributes with either an experiment or a sample."""
+
+    name = models.ForeignKey("OntologyTerm", on_delete=models.CASCADE, related_name="+")
+    unit = models.ForeignKey("OntologyTerm", on_delete=models.CASCADE, related_name="+", null=True)
+    probability = models.FloatField(null=True)
+    source = models.TextField()
+
+    value_type = models.TextField(
+        choices=[
+            (BOOL, "Boolean"),
+            (INT, "Integer"),
+            (FLOAT, "Float"),
+            (ONTOLOGY_TERM, "Ontology Term"),
+        ],
+    )
+    value = models.TextField()
+
+    class Meta:
+        abstract = True
+
+    def to_dict(self):
+        rendered = {
+            "name": self.name.to_dict(),
+            "unit": None if self.unit is None else self.unit.to_dict(),
+            "probability": "unknown" if self.probability is None else self.probability,
+            "source": self.source,
+            "value": self.get_value(),
+        }
+
+        if self.value_type == ONTOLOGY_TERM:
+            rendered["value"] = rendered["value"].to_dict()
+
+        return rendered
+
+    def set_value(self, value):
+        """This method sets the attribute value and assigns the correct
+        value_type. NOTE: we assume that all provided strings are ontology terms."""
```

Comment:
@cgreene is this a good assumption to make? Would we ever want to allow people to just have the value of this metadata be a non-ontology string?

Reply:
What did your design document state? I don't understand "the value of this metadata" - can you give a concrete example so that I can try to answer?

Reply:
About the "value of this metadata", an example of a piece of metadata would be `{'name': 'sex', 'value': 'male'}`. The question is if we always want them to report this value as an ontology term, or if there could be cases where they would report it as a string (maybe for different kinds of metadata).

Reply:
Hmm, I'm pretty sure I imagined these not being strings. I just wanted to make sure we weren't going to box ourselves in here. I can't think of any examples that would really make sense. `{'name': 'name', 'value': 'Kurt Wheeler'}` is the idea, but I don't think we'd ever want to store names... `{'name': 'notes', 'value': 'I sneezed on this sample'}` might also make sense, but I'm not sure where we'd get that kinda data.

Unless you can think of anything that would make sense, I think this is probably fine as is.